src/main/java/dev/jlibra/admissioncontrol/AdmissionControl.java

```diff
@@ -0,0 +1,149 @@
+                .setExpirationTime(transaction.getExpirationTime())
+                .setGasUnitPrice(transaction.getGasUnitPrice())
+                .setMaxGasAmount(transaction.getMaxGasAmount())
+                .setSenderAccount(ByteString.copyFrom(KeyUtils.toByteArrayLibraAddress(publicKey.getEncoded())))
+                .setSequenceNumber(transaction.getSequenceNumber())
+                .build();
+
+        SignedTransaction signedTransaction = SignedTransaction.newBuilder()
+                .setRawTxnBytes(rawTransaction.toByteString())
+                .setSenderPublicKey(ByteString.copyFrom(KeyUtils.stripPublicKeyPrefix(publicKey.getEncoded())))
+                .setSenderSignature(ByteString.copyFrom(LibraHelper.signTransaction(rawTransaction, privateKey)))
+                .build();
+
+        SubmitTransactionRequest submitTransactionRequest = SubmitTransactionRequest.newBuilder()
+                .setSignedTxn(signedTransaction)
+                .build();
+
+        SubmitTransactionResponse response = stub.submitTransaction(submitTransactionRequest);
+
+        channel.shutdown();
+
+        return new SubmitTransactionResult(response.getAcStatus(), response.getMempoolStatus(), response.getVmStatus());
+    }
+
+    private ByteString readCodeFromStream(Transaction transaction) {
+        try {
+            return ByteString.readFrom(transaction.getProgram().getCode());
+        } catch (IOException e) {
+            throw new RuntimeException("Could not read move code from input stream", e);
+        }
+    }
+
+    public UpdateToLatestLedgerResult updateToLatestLedger(List<GetAccountState> arguments) {
+        ManagedChannel channel = ManagedChannelBuilder.forAddress(host, port)
+                .usePlaintext()
+                .build();
+
+        AdmissionControlBlockingStub stub = AdmissionControlGrpc.newBlockingStub(channel);
+
+        List<RequestItem> requestItems = arguments.stream().map(argument -> {
+            GetAccountStateRequest getAccountStateRequest = GetAccountStateRequest.newBuilder()
+                    .setAddress(ByteString.copyFrom(argument.getAddress()))
+                    .build();
+
+            RequestItem requestItem = RequestItem.newBuilder()
+                    .setGetAccountStateRequest(getAccountStateRequest)
+                    .build();
+            return requestItem;
+        }).collect(toList());
+
+        UpdateToLatestLedgerResponse response = stub.updateToLatestLedger(UpdateToLatestLedgerRequest.newBuilder()
+                .addAllRequestedItems(requestItems)
+                .build());
+
+        List<AccountState> accountStates = new ArrayList<>();
+
+        response.getResponseItemsList().forEach(responseItem -> {
+            accountStates.addAll(LibraHelper.readAccountStates(responseItem.getGetAccountStateResponse()
+                    .getAccountStateWithProof()));
+        });
+
+        UpdateToLatestLedgerResult result = UpdateToLatestLedgerResult.create()
```

Comment:
you could return directly without redundant variable

Reply:
This is also fixed now in the other branch I'm working with. I have done some heavy refactoring there so I don't want to cause conflicts by making the same fix here.